### DISCUSSION TITLE
Regarding the refresh issue of ContentControl content in DataGrid

### DISCUSSION THREAD

**Discussion by zg4455** [AvaloniaUI/Avalonia#16207] [Q&A] [answered]
The DataGrid contains the following content

<ContentControl Content="{Binding}" ToolTip.Tip="{Binding FirstName}" ContentTemplate="{x:Static dataTemplates:DataTemplateProvider.GenderDataTemplate}" />

How to achieve automatic refresh of this content? PropertyChanged doesn't seem to work

**Comment by stevemonaco** [comment]
Please provide more details as to how you're using `DataGrid` and the data you're binding to. (ie. show the XAML and the ViewModel code)

**Comment by zg4455** [comment]
[FuncDataTemplateSample.zip](https://github.com/user-attachments/files/16075973/FuncDataTemplateSample.zip)

**Reply by zg4455** [reply]
This is a test instance. I want to refresh the icon on the left

**Reply by zg4455** [reply]
> This is a test instance. I want to refresh the icon on the left

Click the button to refresh the Name, but the icon on the left cannot be refreshed

**Reply by zg4455** [reply]
> > This is a test instance. I want to refresh the icon on the left
> 
> Click the button to refresh the Name, but the icon on the left cannot be refreshed

![image](https://github.com/AvaloniaUI/Avalonia/assets/25166463/1ae764f5-c148-42fa-8dde-ac1a641024c2)

**Comment by stevemonaco** [comment]
So the general topic is about `DataTemplate`, not `DataGrid`, and extending https://github.com/AvaloniaUI/Avalonia.Samples/tree/main/src/Avalonia.Samples/DataTemplates/FuncDataTemplateSample

Understanding how the GUI updates is important here. 1. The `ContentControl` will put the binded content (a `Person`) through the `ContentTemplate` only once. Otherwise the instance itself needs replaced. 2. There's no binding within the `DataTemplateProvider` which could be used to update the generated child control.

The easiest way is to replace the instance with a new one so the new item regenerates the control, IMO. eg.

```cs
private void Bbbb_Click(object? sender, Avalonia.Interactivity.RoutedEventArgs e)
{
    var oldPerson = People[1];

    People[1] = new Person()
    {
        FirstName = DateTime.Now.ToString(),
        LastName = oldPerson.FirstName,
        Sex = Sex.Female
    };
}
```

In this particular case, I might prefer writing two converters if I were writing it from scratch. However, these samples are meant to demonstrate how concepts are used, not necessarily *when* to use them.

**Reply by zg4455** [reply]
In this particular case, I might prefer writing two converters if I were writing it from scratch
Can you explain it in detail

**Reply by zg4455** [reply]
The method of replacing instances will cause changes to the selected row, and I do not want to use this method

**Reply by stevemonaco** [reply, marked as answer]
Just store the selection (`SelectedIndex`) before the change and reassign it afterwards then. It's not a big deal.

On the converter approach, you would write a converter for the Path.Data (since it returns different geometries) and one for Path.Fill (returning different brushes). Then in XAML, something like: `<Path Data="{Binding Sex, Converter={...}}" Fill="{Binding Sex, Converter={...}}" />`. There are many examples on how to write converters, so I'm not covering that.

**Reply by zg4455** [reply]
Thank you very much for your help
